wsmux/session.go

```diff
@@ -0,0 +1,237 @@
+package wsmux
+
+import (
+	"encoding/binary"
+	// "io"
+	"io/ioutil"
+	// "log"
+	"net"
+	// "runtime"
+	"bytes"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	// "github.com/taskcluster/slugid-go/slugid"
+)
+
+const (
+	defaultQueueSize       = 10
+	defaultStreamQueueSize = 10
+)
+
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
+	conn     *websocket.Conn
+
+	acceptDeadline time.Time
+	readDeadline   time.Time
+	writeDeadline  time.Time
+
+	logger Logger
+
+	closed       atomic.Value
```

Comment:
Use a channel instead, call the channel `done`. And say the the channel is closed when underlying connection is closed...

And say inorder to call `close(s.done)` you must hold `s.mu`. That way you can do a non-blocking read from `s.done` to check if it was closed before... As closing a channel twice leads to deadlocks...